modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_temperature_zephyr.c

```diff
@@ -18,66 +18,92 @@
 #include <kernel.h>
 #include <logging/log.h>
 #include <stdbool.h>
+#include <stdint.h>
+#include <zephyr.h>
+
+/** @brief Default temperature [C] reported if NRF_802154_TEMPERATURE_UPDATE is disabled. */
+#define DEFAULT_TEMPERATURE 20
 
 #define LOG_LEVEL LOG_LEVEL_INFO
 #define LOG_MODULE_NAME nrf_802154_temperature
 LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
-/** @brief Value of the last temperature measurement in degree Celsius. */
-static int8_t temperature;
+/** @brief Structure describing the temperature measurement module. */
+static struct {
+	const struct device *dev;	/**< Temperature device pointer. */
+	int8_t value;			/**< Last temperature measurement in degree Celsius. */
 
-void nrf_802154_temperature_init(void)
-{
-	/* Intentionally empty. */
-}
+	struct k_work_delayable work;
+} temperature;
 
-void nrf_802154_temperature_deinit(void)
+#if defined(CONFIG_NRF_802154_TEMPERATURE_UPDATE)
+static void temperature_update_reschedule(void)
 {
-	/* Intentionally empty. */
+	k_work_reschedule(&temperature.work, K_MSEC(CONFIG_NRF_802154_TEMPERATURE_UPDATE_PERIOD));
 }
 
-int8_t nrf_802154_temperature_get(void)
+static void work_handler(struct k_work *work)
 {
-	return temperature;
+	struct sensor_value temperature_value;
+	int err;
+
+	err = sensor_sample_fetch(temperature.dev);
+	if (err < 0) {
+		temperature_update_reschedule();
+		return;
+	}
+
+	err = sensor_channel_get(temperature.dev, SENSOR_CHAN_DIE_TEMP, &temperature_value);
+	if (err < 0) {
+		temperature_update_reschedule();
+		return;
+	}
+
+	if (temperature.value != temperature_value.val1) {
+		temperature.value = temperature_value.val1;
+
+		nrf_802154_temperature_changed();
+	}
+
+	temperature_update_reschedule();
 }
 
-/** @brief Thread handler for temperature update. */
-static void temperature_update_thread(void *p1, void *p2, void *p3)
+static int temperature_update_init(const struct device *dev)
 {
-	ARG_UNUSED(p1);
-	ARG_UNUSED(p2);
-	ARG_UNUSED(p3);
+	ARG_UNUSED(dev);
 
-	static struct sensor_value  temperature_value;
-	static const struct device *temperature_dev;
+	temperature.dev = DEVICE_DT_GET(DT_NODELABEL(temp));
 
-	temperature_dev = DEVICE_DT_GET(DT_NODELABEL(temp));
+	__ASSERT_NO_MSG(device_is_ready(temperature.dev));
 
-	if (!device_is_ready(temperature_dev)) {
-		LOG_ERR("Setup of temperature sensor for nRF 802.15.4 driver failed");
-		__ASSERT_NO_MSG(false);
-	}
+	k_work_init_delayable(&temperature.work, work_handler);
+	k_work_schedule(&temperature.work, K_NO_WAIT);
 
-	while (true) {
-		sensor_sample_fetch(temperature_dev);
-		sensor_channel_get(temperature_dev, SENSOR_CHAN_DIE_TEMP, &temperature_value);
+	return 0;
+}
+#endif /* defined(CONFIG_NRF_802154_TEMPERATURE_UPDATE) */
 
-		if (temperature != temperature_value.val1) {
-			temperature = temperature_value.val1;
+void nrf_802154_temperature_init(void)
+{
+#if defined(CONFIG_NRF_802154_TEMPERATURE_UPDATE)
```

Comment:
this should be handled in header - there is no point making a call if this does not exist